### DISCUSSION TITLE
Adding Device and reflecting on front end

### DISCUSSION THREAD

**Discussion by ShreyanshMalewar** [pyscada/PyScada#188] [Q&A] [answered]
i am using opcua protocol although i realize that there is still configuration that is yet to be done by me, because after adding device, also after creating device property, variables, read tasks i am still unable to understand how to configure a device and test it on front end.
i am using a server for my pyscada and i have pinged the plc on the same server, with static ip

advise from creators is needed

**Comment by clavay** [comment]
Hello,

Can you provide the screenshots of the OPCUA Device configuration and the associated variables ?

Can you provide also the pyscada logs ?
`sudo tail -n1000 /var/log/pyscada/pyscada_debug.log`

**Reply by ShreyanshMalewar** [reply]
hey,

the OPCUA device config is PLC SIMATIC S7-1500, CPU 1515-2 PN
PLC configured with Port 192.168.10.58 
pings on terminal 
![ping](https://github.com/user-attachments/assets/bacf113b-4d88-45a4-8c84-698f2c487828)
![free_opcua_client_ping](https://github.com/user-attachments/assets/a29c1051-8271-4eec-ac7b-3ad3487f9743)
[[04Dec2024 005739] INFO [pyscada.pdf](https://github.com/user-attachments/files/18004240/04Dec2024.005739.INFO.pyscada.pdf)
![IMG_20241203_134934](https://github.com/user-attachments/assets/65232ffe-e1fa-474a-ba86-65569019385f)

**Reply by clavay** [reply]
What is this link for http://114.143.53.27:9207/ ?
I cannot access any webpage here.

Can you send me a screenshot of the pyscada admin interface of the OPCUA Device configuration and the associated variables ?

**Reply by ShreyanshMalewar** [reply]
the link in particular is where my project is
i apologize for sharing this to you without considering that this port won't open internationally

![device1](https://github.com/user-attachments/assets/5da9dd3d-9f19-4334-8630-f5cb0a89611b)
![device2](https://github.com/user-attachments/assets/5db0ba53-25ee-42f6-86b5-1e9cce19670b)

**Reply by clavay** [reply]
Why did you choose an Handler for this device ?
What is the code of your Handler ?

According to the pyscada logs, your handler is not found.

**Reply by ShreyanshMalewar** [reply]
i have removed the handler, i thought Modbus has a handler so an OPCA device will need one as well thus i added it,
after you comment, i have removed it.

question: i need to set the Variable name in the pyscada similar to what is in my PLC device, right?
I have just realized this mistake.
![var2](https://github.com/user-attachments/assets/f6bedcff-12e1-4584-a644-3e7dc1c3022a)
![var1](https://github.com/user-attachments/assets/34128f51-5104-4ca4-93dd-847895234a51)
![IMG-20241205-WA0000](https://github.com/user-attachments/assets/7e803e79-6a57-4e24-9901-a6939ed03fa6)

and for me to check if the device is working, i need to check the variable state? right?
Do i also need to create a OPCUA server on plc side?

**Reply by clavay** [reply]
The variable name in pyscada doesn't matter for the opcua communication.

Can you send me the new logs you have after removing the handler ?

**Reply by ShreyanshMalewar** [reply]
sure i can 
[log_2.txt](https://github.com/user-attachments/files/18032299/log_2.txt)

**Reply by clavay** [reply]
Can you reinstall the pyscada-opcua plugin from this branch : 
https://github.com/pyscada/PyScada-OPCUA/tree/asyncError

Then restart pyscada and send me the logs.

**Reply by ShreyanshMalewar** [reply]
i did what you asked and here are the results 
[log3.txt](https://github.com/user-attachments/files/18037318/log3.txt)

**Reply by ShreyanshMalewar** [reply]
hey if you don't mind can we do a Google meet or a zoom meet.

**Reply by clavay** [reply]
I seems in the logs that you are still using the master branch of the opcua plugin and not the asyncError branch.

Maybe next week for the call.

**Reply by ShreyanshMalewar** [reply]
Actually I did clone the latest changes although they were not being reflected so I actually just updated the __init.py__ file

And sure let me know whenever you're comfortable to arrange a call next week.

**Reply by clavay** [reply]
What says the log after changing the init file ?

**Reply by ShreyanshMalewar** [reply]
i added the plugin once again
here are the logs 
[log4.txt](https://github.com/user-attachments/files/18074888/log4.txt)

**Reply by clavay** [reply]
As you can see [here](https://github.com/pyscada/PyScada-OPCUA/compare/main...asyncError) the line 52 in this branch is `self.url += ":"` but your log refer to the line 52 as:
```
File "/home/pyscada/.venv/lib/python3.8/site-packages/pyscada/opcua/devices/__init__.py", line 52, in _connect
    self._device.opcuadevice.protocol_choices[
```
So my conclusion is that you are not using the `asyncError` branch of the pyscada-opcua plugin.

**Reply by ShreyanshMalewar** [reply]
I checked and I had not uninstalled the OPCA-plugin the right way, I went through the documentation and realized it,
sincere apologies,
let me try again, i will make sure this time i am using the async error branch

**Reply by ShreyanshMalewar** [reply]
hey
i was able to establish a connection, i made the timeout from 10 seconds to 60 seconds in 
        self.inst = Client(url=self.url, timeout=60)
also there is some issue is ns and i value which i need to input correctly in terms of PLC. 
[logs5.txt](https://github.com/user-attachments/files/18091287/logs5.txt)

**Reply by clavay** [reply]
Can you update to the last commit of the `asyncError` branch to catch `CancelledError` exceptions

**Reply by ShreyanshMalewar** [reply]
i updated and here are the results
i have a bit of confusion on what to put for ns and i values as there are multiple 
![opcua-client](https://github.com/user-attachments/assets/ea87fa26-5910-4334-a913-8d7893ad2b2a)


here are the log changes after the commit 
[log6.txt](https://github.com/user-attachments/files/18110212/log6.txt)

**Reply by clavay** [reply]
In your `log6.txt` you are not using the last commit.
Your log shows an error reporting the line 143 different of the [state of the branch](https://github.com/pyscada/PyScada-OPCUA/blob/asyncError/pyscada/opcua/devices/__init__.py#L143)

**Reply by clavay** [reply]
I don't know for the `ns` and `i`. I don't have a OPC-UA instrument to test.

**Reply by ShreyanshMalewar** [reply]
![Screenshot from 2024-12-13 09-48-14](https://github.com/user-attachments/assets/95fc126c-cb07-4cd6-b506-5abb149771d4)

Actually instead of a commit, I edited the file manually, 
I will try to put different ns and i values and update you like wise.
[log7.txt](https://github.com/user-attachments/files/18123502/log7.txt)
.
[log8.txt](https://github.com/user-attachments/files/18123566/log8.txt)

**Reply by clavay** [reply]
I cannot help you if you don't have the same `__init__.py` file as the one I provided in the `asyncError` branch.
Your `log8.txt` has:
```
  File "/home/pyscada/.venv/lib/python3.8/site-packages/pyscada/opcua/devices/__init__.py", line 143, in _read_data
    value = await self.inst.get_node(ns_i).read_value()
```
So your line 143 is not the same as the `asyncError` [branch](https://github.com/pyscada/PyScada-OPCUA/blob/asyncError/pyscada/opcua/devices/__init__.py#L143):
`        ns_i += str(variable.opcuavariable.NamespaceIndex)`

Before posting here, please check the error, and verify if the line in error correspond to the `asyncError` branch.

**Reply by ShreyanshMalewar** [reply]
i apologize for not cross-checking, i think the server did not restart, i will send you the correct logs the next time.

![Screenshot from 2024-12-14 17-27-25](https://github.com/user-attachments/assets/0516dd78-55d4-4eb6-88d1-54fcd3791142)

i will update you soon.

**Reply by ShreyanshMalewar** [reply]
here is an update 
[log9.txt](https://github.com/user-attachments/files/18209195/log9.txt)

if this helps:https://www.blackbox.ai/share/06ad9596-8c3c-43f3-bb77-cb4b83fcbce5

**Reply by clavay** [reply]
Ok, please test the last commit of the asyncError branch.

As I don't have OPCUA device with me, I cannot test it...

**Reply by ShreyanshMalewar** [reply]
These are the updates after latest commit you gave 
[log10.txt](https://github.com/user-attachments/files/18228982/log10.txt)

don't mind me asking, can we schedule a Google meet in the upcoming week?

**Reply by clavay** [reply]
I update the [asyncError](https://github.com/pyscada/PyScada-OPCUA/blob/asyncError/pyscada/opcua/devices/__init__.py#L143) branch and made it work with this [server example](https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/server-minimal.py).
I set the user and password optional.
Please update PyScada-OPCUA, don't forget to run the migrations (migrate command).

Sorry for the meeting I don't have enough time currently.

**Reply by ShreyanshMalewar** [reply]
I thank you, @clavay, for always responding and helping.
With the updated async error branch, I was able to work with a PLC device.
next i will try to work with pyscada hmi section
[log11.txt](https://github.com/user-attachments/files/18419732/log11.txt)
[PLC_commune.pdf](https://github.com/user-attachments/files/18419733/PLC_commune.pdf)

**Reply by clavay** [reply]
If you want to make a tutorial from your PDF, we could add it to the PyScada-OPCUA plugin.

**Reply by ShreyanshMalewar** [reply]
if that PDF helps, please add it to the PyScada-OPCUA plugin,
kindly let me know if you want me to add any more details before finalizing the PDF.

**Reply by clavay** [reply]
I was thinking more about a tutorial with details on how do you connect and validate the values you get in PyScada.

**Reply by ShreyanshMalewar** [reply]
can you please tell me how the tutorial will be, like what specific points should I cover for establishing connection and validating values.
i will upload a Word file so that you can make changes, and then we will finalize a PDF, once you give it a green flag.

**Reply by clavay** [reply]
I was thinking about some text to explain your steps and pictures.

- Which instruments are you connecting and which data will you read.
- How do you configure them in PyScada.
- What is the result you see in the admin and user interfaces.

**Reply by ShreyanshMalewar** [reply]
[Connect & Validate PLC.pdf](https://github.com/user-attachments/files/18705932/Connect.Validate.PLC.pdf)

**Reply by clavay** [reply, marked as answer]
Thank you for the tutorial, I will test it as we have siemens SIMATIC S7-1200.
After that we can write this tutorial as an _rst_ file and add it to the [documentation](https://pyscada.readthedocs.io/).

**Comment by clavay** [comment]
Good news.
Your logs seems to be duplicated, look at the config in `settings.py`.

One question, I see your device read 6 values for each variable (one value don't pass the COV check, so 5 are kept) each pooling interval and the default handler shoud read only 1 value: Did you made a specific device handler for your OPC-UA device ?

I close the discussion.
Please open an issue in the corresponding project if needed (PyScada, PyScada-OPCUA,...).

**Reply by ShreyanshMalewar** [reply]
as per your advice, i have never used any device handler for the OPC-UA device
i am not sure as to why i am getting 6 values, it can be due to me setting the values in the write task as 6 and more?
and i kept the pooling interval as 5 seconds while creating the device maybe that's why
![write task](https://github.com/user-attachments/assets/a09dac99-b771-4df1-abc1-a98228ad916a)

i will run few more tests and get back to you with a proper reason.
